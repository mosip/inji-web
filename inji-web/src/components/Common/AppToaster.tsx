import React from "react";
import {ToastContainer} from "react-toastify";

export const AppToaster: React.FC = () => {
    return <ToastContainer
        position="top-right"
        autoClose={1000}
        hideProgressBar
        newestOnTop
        closeOnClick
        rtl={false}
        icon={<React.Fragment/>}
        pauseOnFocusLoss
        draggable
        pauseOnHover
        style={{width: '400px'}}
        theme="colored"
    />
}
