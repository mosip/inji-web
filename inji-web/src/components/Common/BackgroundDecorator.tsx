import React from 'react';

interface BackgroundDecoratorProps {
    circleCount?: number;
    topOffset?: number;
    baseRadius?: number;
    logoSrc?: string;
    logoAlt?: string;
    testId?: string;
}

export const BackgroundDecorator: React.FC<BackgroundDecoratorProps> = ({
    circleCount = 6,
    topOffset = 155,
    baseRadius = 96,
    logoSrc,
    logoAlt = "Logo",
    testId = "background-decorator"
}) => {
    return (
        <div 
            className="overflow-hidden absolute inset-0 z-0 flex items-center justify-center pointer-events-none"
            data-testid={testId}
        >
            <div className={`absolute top-[${topOffset}px]`} style={{ top: `${topOffset}px` }}>
                {[...Array(circleCount)].map((_, index) => {
                    const radius = baseRadius + index * baseRadius;
                    const opacity = 0.8 - index * 0.1;
                    return (
                        <div
                            key={index}
                            className="absolute rounded-full border overflow-hidden"
                            style={{
                                width: `${radius}px`,
                                height: `${radius}px`,
                                borderWidth: '1px',
                                borderColor: `rgba(228, 231, 236, ${opacity})`,
                                top: `calc(50% - ${radius / 2}px)`,
                                left: `calc(50% - ${radius / 2}px)`
                            }}
                            data-testid={`${testId}-circle-${index}`}
                        />
                    );
                })}
                
                {logoSrc && (
                    <div
                        className="flex items-center justify-center"
                        data-testid={`logo-${testId}-container`}
                    >
                        <img
                            src={logoSrc}
                            alt={logoAlt}
                            data-testid={`logo-${testId}`}
                        />
                    </div>
                )}
            </div>
        </div>
    );
};