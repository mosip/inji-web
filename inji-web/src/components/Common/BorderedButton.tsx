import React from "react";

export const BorderedButton:React.FC<BorderedButtonProps> = (props) => {
  return <div className={"bg-gradient-to-r rounded-lg from-iw-primary to-iw-secondary p-0.5"}>
            <div className="py-2 px-4 rounded-lg bg-white">
                <button
                    data-testid={props.testId}
                    onClick={props.onClick}
                    className="text-iw-tertiary font-bold rounded-lg">
                    {props.title}
                </button>
            </div>
    </div>
}

export type BorderedButtonProps = {
    testId: string;
    onClick: ()=> void;
    title: string;
}
