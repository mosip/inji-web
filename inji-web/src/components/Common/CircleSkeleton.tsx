type CircleSkeletonProps = {
    size?: string;
    className?: string;
};

export const CircleSkeleton = ({
                                   size = 'w-20 h-20',
                                   className = '',
                               }: CircleSkeletonProps) => {
    return (
        <div
            className={`rounded-full bg-gray-300 animate-pulse ${size} ${className}`}
        />
    );
};
  