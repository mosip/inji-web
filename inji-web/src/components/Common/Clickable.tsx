import React from 'react';

export interface ClickableProps {
    onClick: () => void;
    children: React.ReactNode;
    testId: string;
    className: string;
}

export const Clickable: React.FC<ClickableProps> = ({
                                                        onClick,
                                                        children,
                                                        testId,
                                                        className = '',
                                                    }) => {
    const handleKeyPress = (event: React.KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'Enter' || event.key === ' ') {
            event.preventDefault();
            onClick();
        }
    };

    return (
        <div
            role="menuitem"
            tabIndex={0}
            className={className}
            onClick={onClick}
            onKeyDown={handleKeyPress}
            data-testid={testId}
        >
            {children}
        </div>
    );
};
