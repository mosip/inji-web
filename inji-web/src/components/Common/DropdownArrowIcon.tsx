import React from 'react';
import {RiArrowUpSFill, RiArrowDownSFill} from 'react-icons/ri';
import { DropdownArrowIconProps } from '../../types/data';
import { GradientWrapper } from './GradientWrapper';

const DropdownArrowIcon: React.FC<DropdownArrowIconProps> = ({isOpen, size = 20, color = 'var(--iw-color-languageArrowIcon)'}) => {
    const ArrowIcon = isOpen ? RiArrowUpSFill : RiArrowDownSFill;

    return (
        <GradientWrapper>
            <ArrowIcon size={size} color={color} />
        </GradientWrapper>
    );
};

export default DropdownArrowIcon;
