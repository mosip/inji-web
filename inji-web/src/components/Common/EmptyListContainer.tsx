import React from "react";
import {EmptyListContainerProps} from "../../types/components";

export const EmptyListContainer: React.FC<EmptyListContainerProps> = ({content}) => {
    return <React.Fragment>
        <div data-testid="EmptyList-Outer-Container"
             className="flex justify-center items-center w-full mx-auto flex-col">
            <div data-testid="EmptyList-Inner-Container" className="container mx-auto mt-8 px-4 flex-1 flex flex-col">
                <p data-testid="EmptyList-Text" className="text-center">{content}</p>
            </div>
        </div>
    </React.Fragment>
}
