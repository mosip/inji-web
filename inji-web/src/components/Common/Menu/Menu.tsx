import React, {useEffect, useRef, useState} from "react";
import {MenuItem} from "./MenuItem";
import {MenuItemType} from "../../../types/data";
import {MenuStyles} from "./MenuStyles";

interface MenuProps {
    triggerComponent: React.ReactNode;
    menuItems: MenuItemType[];
    testId: string;
}

export const Menu: React.FC<MenuProps> = ({triggerComponent, menuItems, testId}) => {
    const [isOpen, setIsOpen] = useState(false);
    const buttonRef = useRef<HTMLButtonElement>(null);
    const menuRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        const handleClickOutside = (event: MouseEvent) => {
            if (
                menuRef.current &&
                !menuRef.current.contains(event.target as Node) &&
                buttonRef.current &&
                !buttonRef.current.contains(event.target as Node)
            ) {
                setIsOpen(false);
            }
        };

        if (isOpen) {
            document.addEventListener("mousedown", handleClickOutside);
        }

        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, [isOpen]);


    return (
        <div className={MenuStyles.menu.container} data-testid={`${testId}-menu`}>
            <button
                className={MenuStyles.menu.button}
                ref={buttonRef}
                onClick={(event) => {
                    event.stopPropagation();
                    setIsOpen((prev) => !prev);
                }}
                data-testid={`icon-${testId}-menu`}
            >
                {triggerComponent}
            </button>

            {isOpen && (
                <div
                    ref={menuRef}
                    role="menu"
                    data-testid={`menu-${testId}`}
                    className={MenuStyles.menu.menuContainer}
                    style={{zIndex: 9999}}
                >
                    <div className={MenuStyles.menu.caretOuter}></div>
                    <div className={MenuStyles.menu.caretInner}></div>

                    {menuItems.map((item) => (
                        <MenuItem
                            key={item.id}
                            label={item.label}
                            icon={item.icon}
                            color={item.color}
                            onClick={() => {
                                item.onClick();
                                setIsOpen(false);
                            }}
                            testId={item.id}
                        />
                    ))}
                </div>
            )}
        </div>
    );
};
