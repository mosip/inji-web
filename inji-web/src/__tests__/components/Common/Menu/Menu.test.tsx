import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import {Menu} from "../../../../components/Common/Menu/Menu";
import {Clickable} from "../../../../components/Common/Clickable";

describe('EllipsisMenu Component', () => {
    const mockMenuItems = [
        {label: 'Edit', onClick: jest.fn(), id: "edit"},
        {label: 'Delete', onClick: jest.fn(), id: "delete"}
    ];

    beforeEach(() => {
        jest.clearAllMocks();
    });

    const triggerComponent = <span data-testid="trigger-menu">...</span>;

    it('renders correctly and matches snapshot', () => {
        const {asFragment} = render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems}
                                          testId={"test-view"}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the ellipsis button', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);

        const button = screen.getByTestId('trigger-menu');

        expect(button).toBeInTheDocument();
    });

    it('does not show menu items when initially rendered', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);

        expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    });

    it('shows menu items when button is clicked', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);

        const button = screen.getByTestId('trigger-menu');
        fireEvent.click(button);

        const menu = screen.getByRole('menu');
        expect(menu).toBeInTheDocument();

        const menuItems = screen.getAllByRole('menuitem');
        expect(menuItems).toHaveLength(2);
        expect(menuItems[0]).toHaveTextContent('Edit');
        expect(menuItems[1]).toHaveTextContent('Delete');
    });

    it('calls the onClick handler and closes menu when a menu item is clicked', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);

        const button = screen.getByTestId('trigger-menu');
        fireEvent.click(button);

        const editMenuItem = screen.getByTestId('menu-item-edit');
        fireEvent.click(editMenuItem);

        expect(mockMenuItems[0].onClick).toHaveBeenCalledTimes(1);
        // Check if the menu is closed after clicking an item
        expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    });

    it('toggles menu visibility when button is clicked multiple times', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);

        const button = screen.getByTestId('trigger-menu');

        // Initially not visible
        expect(screen.queryByRole('menu')).not.toBeInTheDocument();
        // First click - show menu
        fireEvent.click(button);
        expect(screen.getByRole('menu')).toBeInTheDocument();
        // Second click - hide menu
        fireEvent.click(button);
        expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    });

    it('stops event propagation when clicking the button', () => {
        const mockContainerClick = jest.fn();

        render(
            <Clickable onClick={mockContainerClick} testId={'test-view-wrapper'}>
                <Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>
            </Clickable>
        );

        const button = screen.getByTestId('trigger-menu');
        fireEvent.click(button);

        expect(mockContainerClick).not.toHaveBeenCalled();
    });

    it('closes the menu when clicking outside', () => {
        render(
            <div>
                <div data-testid="outside-element">Outside</div>
                <Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>
            </div>
        );

        // Open the menu
        const button = screen.getByTestId('trigger-menu');
        fireEvent.click(button);
        // Menu should be visible
        expect(screen.getByRole('menu')).toBeInTheDocument();
        // Click outside the menu
        const outsideElement = screen.getByTestId('outside-element');
        fireEvent.mouseDown(outsideElement);

        // Menu should be closed
        expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    });

    it('does not close the menu when clicking inside the menu', () => {
        render(<Menu triggerComponent={triggerComponent} menuItems={mockMenuItems} testId={"test-view"}/>);
        // Open the menu
        const button = screen.getByTestId('trigger-menu');
        fireEvent.click(button);
        // Click on the menu container (not on a menu item)
        const menu = screen.getByRole('menu');
        fireEvent.mouseDown(menu);

        // Menu should still be visible
        expect(screen.getByRole('menu')).toBeInTheDocument();
    });
});