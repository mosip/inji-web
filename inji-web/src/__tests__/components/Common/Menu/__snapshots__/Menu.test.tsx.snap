// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EllipsisMenu Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block text-left"
    data-testid="test-view-menu"
  >
    <button
      class="p-2 hover:bg-gray-100"
      data-testid="icon-test-view-menu"
    >
      <span
        data-testid="trigger-menu"
      >
        ...
      </span>
    </button>
  </div>
</DocumentFragment>
`;
